### DISCUSSION TITLE
如何处理多个baseapp进程间同步问题（basepp负载均衡）

### DISCUSSION THREAD

**Discussion by Leovany** [kbengine/kbengine#1432] [Q&A]
开发期间，一直是单baseapp开发，现在考虑到人数增加问题，需要搞baseapp负载均衡，但相关资料较少，向大家请教一下
目前，有个定义了个全局实体类`GameCenter`, 将全局相关的功能都封装在这个类上
里面包含两种情况，
第一种：记录在线玩家的字典（直接挂在GameCenter上），
第二种：以及全局功能管理类（比如排行榜，活动等等，一个对象xxxManger挂在GameCenter上），

当在`start_server.bat`，增加个baseapp进程后，客户端连接第一个baseapp 正常，但如果连到第二个baseapp 就报错了
看了下log, 发现了几个问题
1.  baseapp 上的字典列表等结构 、子对象功能模块（xxxManger），并没有同步到第二个baseapp上
2.  baseapp之间调用的方法，是必须在def 文件中声明吗？
3. GameCenter实体上可以在def文件声明，但子对象 中的方法（比如`RankManger.getRank()`）该如何声明？
4. 因为`GameCenter`挂着很多功能模块，似乎 `KBEngine.globalData` 会同步很多数据，是否将每个模块拆开成单独一个实体类，这样子更好？

PS :  主要单个baseapp会写，但多个baseapp 的注意事项，以及写法 不清楚，官方文档上似乎这一块没讲的特别清楚，希望有大佬可以帮忙捋一捋，或者有相关资料，demo 可以参考学习一下，十分感谢！

附截图
1. `GameCenter`.cs
![image](https://user-images.githubusercontent.com/2864116/118772811-bda1fa80-b8b6-11eb-85c4-b9e4c5753977.png)

2. Account.cs 
![image](https://user-images.githubusercontent.com/2864116/118773048-fcd04b80-b8b6-11eb-9000-4cc503af3d79.png)

3. base目录下的kbemain.cs
> `GameCenter` , 两个baseapp 之间没同步, 获取不到
![image](https://user-images.githubusercontent.com/2864116/118774514-80d70300-b8b8-11eb-9d4c-71a1be7c8df2.png)

![image](https://user-images.githubusercontent.com/2864116/118774596-99471d80-b8b8-11eb-99e9-3e1dde27769b.png)

4. GameCenter 上 子对象方法获取不到
比如下图，
serverVersionManager 是挂在GameCenter上，
玩家通过 `KBEngine.globalData["GameCenter"].getManager(gameManager.ManagerConst.ServerVersion)` 获取 `serverVersionManager `,然后调用里面的方法`getLatestVersion`, 提示方法不存在，
然后在 `GameCenter`中`BaseMethods`,  def 定义 `getLatestVersion` ,也不行，提示GameCenter没有这个方法
目前我想到的一个方案是，将把各个manger  , 都单独拆开成一个实体类（跟GameCenter一样），改动就比较大了
像我这这种情况怎么处理比较好？ 
![image](https://user-images.githubusercontent.com/2864116/118776304-6dc53280-b8ba-11eb-95b5-5bb1ad452903.png)

5. globalData
像下面这样的需求，要怎么弄比较好？ 
![image](https://user-images.githubusercontent.com/2864116/118778638-e1683f00-b8bc-11eb-9d74-dd6816d2ee5f.png)

**Comment by kaisong524118** [comment]
base2 访问 base1的实时时它只是一个 ”entityCall"  是不能通过 getxxx() : return obj 方法返回对象的

**Reply by Leovany** [reply]
> base2 访问 base1的实时时它只是一个 ”entityCall" 是不能通过 getxxx() : return obj 方法返回对象的

啊这....其实我需求是 全局GameCenter 就是一份数据，多个baseapp访问的是同一份数据（结构来说，这些似乎放到baseappmgr ）， 那怎么处理比较好？ 

感谢您的回复

**Reply by kaisong524118** [reply]
其实我也在纠结这个问题，不知道要怎样设计才合理些！

**Reply by Leovany** [reply]
哎，kbe资料比较少，头疼

**Comment by overwhelm** [comment]
直接用globalData ，有数据变化时，读出数据，修改数据后，重新赋值呢？

**Reply by overwhelm** [reply]
或者用redis来存储数据

**Reply by Leovany** [reply]
> globalData

你好，我现在用的就是 globalData 方式，但他没有同步到其他baseapp，
就是是文档API 说的情况，顶层才会广播，但文档没有说该怎么做才好
其次，我问一下，每次重新赋值，效率影响大吗，这个还挺频繁的修改

感谢你的回复

**Reply by overwhelm** [reply]
应该不太影响吧，就相当于变量的赋值了
